infoschema/cache.go

```diff
@@ -98,48 +124,53 @@ func (h *InfoCache) GetBySnapshotTS(snapshotTS uint64) InfoSchema {
 	defer h.mu.RUnlock()
 
 	infoschema_metrics.GetTSCounter.Inc()
-	if snapshotTS >= h.maxUpdatedSnapshotTS {
-		if len(h.cache) > 0 {
-			infoschema_metrics.HitTSCounter.Inc()
-			return h.cache[0]
-		}
+	if schema, err := h.getSchemaByTimestampNoLock(snapshotTS); err == nil {
+		infoschema_metrics.HitTSCounter.Inc()
+		return schema
 	}
 	return nil
 }
 
 // Insert will **TRY** to insert the infoschema into the cache.
 // It only promised to cache the newest infoschema.
 // It returns 'true' if it is cached, 'false' otherwise.
-func (h *InfoCache) Insert(is InfoSchema, snapshotTS uint64) bool {
+// snapshotTS is the timestap of the schema update, schemaTS is the timestamp of the schema taking effect
+func (h *InfoCache) Insert(is InfoSchema, snapshotTS, schemaTS uint64) bool {
```

Comment:
We can remove `snapshotTS` now. It is not used anywhere.

Change comments of `GetBySnapshotTS`, too: `snapshotTS` was `startTS`. But in `GetBySnapshotTS`, we mean that, give any random TS, we should return the correct schema or not.